feat(ffe-datepicker-react): Make calendar readable for screenreaders

henninghuseby opened the pull request:
## Beskrivelse

I denne PR-en har input og kalenderknappen blitt skilt fra hverandre, slik at kalenderen kun åpnes når kalderknappen trykkes på. Det har blitt lagt på focus-trap i kalenderen når den er åpen, slik at det er mulig å tabbe seg fra datoene i kalenderen til neste/forrige mnd knappenen i toppen av kalenderen. Fokus-styling har også blitt oppdatert slik at det skal bli lettere å se hvilket element som har fokus. Hver dato har fått lagt aria-label slik at skjermlesere lettere skjønner hvilken dato de har fokus på. Kalenderknappen har fått label som oppdaterer seg i forhold til om dato er valgt eller ikke

## Motivasjon og kontekst

Datovelgeren er vanskelig å bruke for de meg skjermleser, dette er også dokumentert i Funka sin rapport 

## Testing
Kjørt npm start og sjekket endringen på localhost. Sjekket med skjermleser at input feltet blir opples som forventet og navigeringen i kalenderen oppfører seg som forventet. Tab-knappen flytter fokus mellom dato-tabellen og frem/tilbake knappene, når dato-tabellen har fokus kan en navigere rundt i tabellen med piltastene, esc-knappen lukker kalenderen og setter fokus tilbake til kalender-knappen. 
Oppdatert tester i fohold til ny funksjonalitet

torkjels commented on packages/ffe-datepicker-react/src/calendar/ActiveDate.js:
```diff
@@ -46,13 +62,18 @@ export default class ActiveDate extends Component {
                 className="ffe-calendar__day"
                 headers={headers}
                 onClick={() => onClick(date)}
```
Burde denne onClick'en heller vært på knappen inni `td`en? Og da kunne `tabIndex` blitt fjernet fra `td` og heller blitt satt på `button`?

torkjels commented on packages/ffe-datepicker-react/src/calendar/ActiveDate.js:
```diff
@@ -46,13 +62,18 @@ export default class ActiveDate extends Component {
-                }}
+                ref={activeRef}
                 role="gridcell"
+                date-number={date.date}
```
Hva gjør denne? Den ser ikke ut som en gyldig attributt :thinking:

henninghuseby replied:
Henger igjen fra litt testing, er are en custom html attributt som en kan slenge på hvor som helst egentlig
Fjerner den jeg :D

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -166,13 +166,27 @@ export default class Calendar extends Component {
+            document.activeElement !== this.nextMonthButtonElementRef.current &&
+            document.activeElement !== this.prevMonthButtonElementRef.current
+                ? true
+                : false;
```
Denne kan forkortes til
```
const forceFocus =
            document.activeElement !== this.nextMonthButtonElementRef.current &&
            document.activeElement !== this.prevMonthButtonElementRef.current;
```

torkjels commented on packages/ffe-datepicker-react/src/calendar/Header.js:
```diff
@@ -58,4 +62,6 @@ Header.propTypes = {
     previousMonthLabel: string.isRequired,
     year: number.isRequired,
+    prevMonthButtonElement: object,
+    nextMonthButtonElement: object,
```
Slenge på en `.isRequired` på disse to?

torkjels commented on packages/ffe-datepicker-react/src/datepicker/Datepicker.js:
```diff
@@ -67,25 +64,19 @@ export default class Datepicker extends Component {
+        if (
+            prevProps.value !== this.props.value &&
+            prevState.displayDatePicker &&
+            !this.state.displayDatePicker
```
Denne er litt vanskelig å henge med på. Hadde det vært mulig å dra den ut i en variabel med et beskrivende navn? Eller?

torkjels commented on packages/ffe-datepicker-react/src/datepicker/Datepicker.js:
```diff
@@ -67,25 +64,19 @@ export default class Datepicker extends Component {
-            });
-            return;
-        }
         this.validateDateIntervals();
```
Trengs denne lenger? Det skal vel ikke bare valideres på fokus?

torkjels commented on packages/ffe-datepicker-react/src/datepicker/Datepicker.js:
```diff
@@ -348,27 +309,37 @@ export default class Datepicker extends Component {
+                            ref={c => {
+                                this._button = c;
+                            }}
+                            onClick={this.clickHandler}
```
Denne kunne gjerne ha blitt kalt feks `calendarButtonClickHandler` eller noe. Eventuelt kan jo alt som har med kalender-knappen å gjøre bli dratt ut i en egen komponent slik som `DateInput`?

torkjels commented on packages/ffe-datepicker-react/src/i18n/nn.js:
```diff
@@ -41,10 +41,15 @@ export default {
     MAX_DATE: 'Oppgitt dato er for langt frem',
     CHOOSE_DATE: 'Velg dato',
+    CHANGE_DATE: 'Endre dato',
+    CHOSEN_DATE: 'valgt dato er',
```
Nå skal jeg ikke si for sikkert her, men jeg tror det er "valt" på nynorsk?

torkjels commented:
Commitene kan kanskje bli squashet til én? Med mindre du føler de kan splittes opp i to: først én for refaktorering/rydding og så én for fiksing. Men uansett tror jeg du trenger å nevne noe om breaking change, ref https://github.com/SpareBank1/designsystem/wiki/Bidra:-Fra-issue-til-pull-request#commit-meldinger

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -19,12 +19,10 @@ export default class Calendar extends Component {
                 props.maxDate,
                 props.language,
             ),
+            isCalendarBlured: false,
```
Kunne denne hett noe sånt som `isFocusingHeader`? Headeren er strengt tatt en del av `Calendar`-komponenten, så `isCalendarBlurred` blir litt misvisende.
Da kunne vi også kalt `isBlured` i `ActiveDate` for `isFocusingHeader`, så blir det enklere å forstå inni `ActiveHeader` hva den prop'en representerer.

torkjels commented on packages/ffe-datepicker-react/src/calendar/ActiveDate.js:
```diff
@@ -1,6 +1,16 @@
+import moment from 'moment';
+import i18n from '../i18n/i18n';
 
 export default class ActiveDate extends Component {
```
Jeg mener å huske at vi kom fra til at `ActiveDate` er en hvilken som helst klikkbar dato, ikke bare den valgte datoen. Og `LeadDate` er ikke-klikkbare datoer. Skulle vi ha endra på denne navngivinga mens vi først er i gang? Så slipper nestemann å slite med det samme. Og i samme slengen endra `isLead` i `simplecalendar.js`. Feks `ActiveDate` til `ClickableDate` og `LeadDate` til `NonclickableDate` eller noe. Og da `isLead` til `isClickable` eller `isThisMonth` eller `isInCurrentMonth` eller `isPartOfCurrentMonth` eller noe.

Det er forsåvidt sånt som godt kan være del av en egen refactor-commit først, og så kommer den faktiske feature-endringen i en commit nr. 2, så blir det enklere å holde styr på. Syns hvertfall jeg :)

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -166,13 +166,27 @@ export default class Calendar extends Component {
+                activeRef={
+                    date.isFocus
+                        ? this.activeDateSelectedRef
+                        : this.activeDateRef
```
Det ser ikke ut til at `ActiveDate` gjør noe med `activeRef` med mindre datoen er fokusert. Hadde det vært mulig å ikke sende inn noen ref hvis `date.isFocus` er `false`? Altså kjøre `activeRef={date.isFocus ? this.activeDateSelectedRef : undefined}`. Og så slette `activeDateRef` helt siden det ikke ser ut til at den brukes noe sted.
Og da kan du kanskje kalle `activeDateSelectedRef` for `selectedDateRef` for å få et litt kortere navn.
Og kanskje kalle `activeRef`-propen for `dateButtonRef` eller noe :shrug:

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -163,16 +162,27 @@ export default class Calendar extends Component {
+                forceFocus={forceFocus}
+                language={this.props.language}
+                activeRef={date.isFocus ? this.activeDateRef : undefined}
+                isBlured={this.state.isFocusingHeader}
```
Denne prop'en kan sikkert også kalles for `isFocusingHeader`. `isBlured` er et litt intetsigende navn. Spesielt når vi inni `ClickableDate` har `isFocus && !isBlured`. Da blir det mer forståelig `isFocus && !isFocusingHeader` syns jeg :)

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -163,16 +162,27 @@ export default class Calendar extends Component {
+        const { calendar } = this.state;
+
+        if (date.isNonClickableDate) {
+            return <NonclickableDate key={date.date} date={date} />;
```
Linje 167 har `NonClickable` (med stor C) og linje 168 har `Nonclickable` (med liten c). Usikker på hvilken som blir mest riktig (kanskje `Nonclickable`?), men det viktigste er at vi holder oss til én konvensjon :)

henninghuseby replied:
Tror jeg går for NonClickabel siden det på engelsk enten er non clickabel eller non-clickable

torkjels replied:
`NonClickable` isåfall :wink:

torkjels commented on packages/ffe-datepicker-react/src/calendar/ClickableDate.js:
```diff
@@ -0,0 +1,100 @@
+                aria-selected={date.isSelected}
+                className="ffe-calendar__day"
+                headers={headers}
+                ref={activeRef}
```
Jeg vet ikke helt hva som blir riktig av fokus i en tabell med knapper, men skulle denne ref'en egentlig vært på `button`-elementet? Siden vi programmatisk setter fokus på den, og det er tross alt knappen vi har `onClick` på. Og sånn sett også fjernet `tabIndex={0}` fra `td`.
Det samme gjelder aria-attributtene her. Er det riktig at de skal være på `td` og ikke på den klikkbare `button`?

henninghuseby replied:
Etter å ha sjekket mer her så tror jeg faktisk det er bedre å ha klikk-eventet på `td` og ikke en egen knapp inne i `td`. Dette siden det blir veldig mye nøsting inne tabellen som gjør at skjermlesere leser opp tabellen veldig rart. Har også sjekket andre som har valgt å droppe knapp inn i `td`. 
Alternativet er å lage custom "tabell" med div-er og knapper med tilhørende roller som da gjør det til en tabell

torkjels replied:
Ja det blir kanskje vel mye å lese opp med mye forskjellig nøsting. Klarer vi å få fram for skjermlesere at td'en er klikkbar?

henninghuseby replied:
Siden vi bruker `aria-selected=true/false` så tror jeg det skal være nok.

torkjels replied:
Muligens. Litt usikker på om man egentlig må formidle det vha. en `role`. Men `td` har jo en innebygd `role` jeg ikke tror vi har lyst å overstyre. Men du får teste litt :)

torkjels commented on packages/ffe-datepicker-react/src/calendar/ClickableDate.js:
```diff
@@ -0,0 +1,100 @@
+    onClick: func.isRequired,
+    forceFocus: bool.isRequired,
+    language: string.isRequired,
+    activeRef: object,
```
`activeRef` henger kanskje litt igjen fra da komponenten het `ActiveDate`? Skulle vi heller kalt den feks `dateButtonRef`?

torkjels commented on packages/ffe-datepicker-react/src/calendar/Calendar.js:
```diff
@@ -163,16 +162,27 @@ export default class Calendar extends Component {
                 key={date.date}
                 onClick={this.mouseClick}
-                forceFocus={this.forceDateFocus}
+                forceFocus={forceFocus}
```
Hadde det holdt å bruke `isFocusingHeader` her? Altså hvis vi fjerner `forceFocus` helt (fjern linje 171-173 og linje 182), kaller `isBlured` for `isFocusingHeader` og så kan `ClickableDate` sin `focusIfNeeded` være
```
    focusIfNeeded() {
        const { date, isFocusingHeader, ..... } = this.props;
        if (date.isFocus && !isFocusingHeader) {
            .....
        }
    }
```

kwltrs commented on packages/ffe-datepicker-react/src/button/Button.js:
```diff
@@ -0,0 +1,32 @@
+            <button
+                ref={buttonRef}
+                onClick={onClick}
+                className="ffe-datepicker--button"
```
Klassnavnet `ffe-datepicker--button` strider mot [BEM-konvensjonen ](https://github.com/SpareBank1/designsystem/wiki/Bidra:-Fra-issue-til-pull-request#bem---block-element-modifier) fordi `--button` er en modifier.  Men et element passer bedre, dvs `ffe-datepicker__button`.

kwltrs reviewed (changes requested):
Kan du dele opp det i to commits og ta alle endringer i `ffe-datepicker` som en egen commit? Systemet bak hvordan designsystemet håndterer versjonering ([conventional commits](https://github.com/SpareBank1/designsystem/wiki/Bidra:-Fra-issue-til-pull-request#commit-meldinger)) trenger at kun én pakke per commit blir endret slik at det kan bumpe riktig pakke og oppdatere riktig CHANGELOG.

henninghuseby committed:
feat(ffe-datepicker): Make calendar readable for screenreaders

This commit improves focus, active and selected styling for datepicker

BREAKING CHANGE: Improve usability of datepicker.

This commit separates date-input and calendar-button to improve accessibility

henninghuseby committed:
feat(ffe-datepicker-react): Make calendar readable for screenreaders

BREAKING CHANGE: Improve usability of datepicker.

This commit separates date-input and calendar-button to improve accessibility

Previously, datepicker toggled on focus on input field. Now calendar-button will now toggle the datepicker, and tab-key will move focus between date in datepicker and prev- and next-month buttons. Select date, esc-key and click outside datepicker will close datepicker

This version adds aria label to each date in datpicker to improve readability for screen readers

kwltrs commented on packages/ffe-datepicker/less/dateinput.less:
```diff
@@ -1,3 +1,5 @@
+@import '../node_modules/@sb1/ffe-form/less/input-field';
```
Bedre å importere denne som [reference](http://lesscss.org/features/#import-atrules-feature-reference). Da kan `ffe-input-field` brukes som mixin, men man får ikke hele definisjonen med i css-en.

```css
@import (reference) '../node_modules/@sb1/ffe-form/less/input-field';
```
```
